// @modern-js/runtime@2.60.0/dist/cjs/core/constants.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  ROUTER_DATA_JSON_ID: () => ROUTER_DATA_JSON_ID,
  RenderLevel: () => RenderLevel,
  SSR_DATA_JSON_ID: () => SSR_DATA_JSON_ID
});
module.exports = __toCommonJS(constants_exports);
var RenderLevel;
(function(RenderLevel2) {
  RenderLevel2[RenderLevel2["CLIENT_RENDER"] = 0] = "CLIENT_RENDER";
  RenderLevel2[RenderLevel2["SERVER_PREFETCH"] = 1] = "SERVER_PREFETCH";
  RenderLevel2[RenderLevel2["SERVER_RENDER"] = 2] = "SERVER_RENDER";
})(RenderLevel || (RenderLevel = {}));
const SSR_DATA_JSON_ID = "__MODERN_SSR_DATA__";
const ROUTER_DATA_JSON_ID = "__MODERN_ROUTER_DATA__";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ROUTER_DATA_JSON_ID,
  RenderLevel,
  SSR_DATA_JSON_ID
});
