// @modern-js/runtime@2.60.0/dist/cjs/cli/constants.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  APP_FILE_NAME: () => APP_FILE_NAME,
  ENTRY_BOOTSTRAP_FILE_NAME: () => ENTRY_BOOTSTRAP_FILE_NAME,
  ENTRY_POINT_FILE_NAME: () => ENTRY_POINT_FILE_NAME,
  ENTRY_POINT_REGISTER_FILE_NAME: () => ENTRY_POINT_REGISTER_FILE_NAME,
  ENTRY_POINT_RUNTIME_GLOBAL_CONTEXT_FILE_NAME: () => ENTRY_POINT_RUNTIME_GLOBAL_CONTEXT_FILE_NAME,
  ENTRY_POINT_RUNTIME_REGISTER_FILE_NAME: () => ENTRY_POINT_RUNTIME_REGISTER_FILE_NAME,
  ENTRY_SERVER_BOOTSTRAP_FILE_NAME: () => ENTRY_SERVER_BOOTSTRAP_FILE_NAME,
  INDEX_FILE_NAME: () => INDEX_FILE_NAME,
  SERVER_ENTRY_POINT_FILE_NAME: () => SERVER_ENTRY_POINT_FILE_NAME
});
module.exports = __toCommonJS(constants_exports);
const APP_FILE_NAME = "App";
const ENTRY_POINT_FILE_NAME = "index.jsx";
const SERVER_ENTRY_POINT_FILE_NAME = "index.server.jsx";
const INDEX_FILE_NAME = "index";
const ENTRY_BOOTSTRAP_FILE_NAME = "bootstrap.jsx";
const ENTRY_SERVER_BOOTSTRAP_FILE_NAME = "bootstrap.server.jsx";
const ENTRY_POINT_RUNTIME_REGISTER_FILE_NAME = "runtime-register.js";
const ENTRY_POINT_RUNTIME_GLOBAL_CONTEXT_FILE_NAME = "runtime-global-context.js";
const ENTRY_POINT_REGISTER_FILE_NAME = "register.js";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  APP_FILE_NAME,
  ENTRY_BOOTSTRAP_FILE_NAME,
  ENTRY_POINT_FILE_NAME,
  ENTRY_POINT_REGISTER_FILE_NAME,
  ENTRY_POINT_RUNTIME_GLOBAL_CONTEXT_FILE_NAME,
  ENTRY_POINT_RUNTIME_REGISTER_FILE_NAME,
  ENTRY_SERVER_BOOTSTRAP_FILE_NAME,
  INDEX_FILE_NAME,
  SERVER_ENTRY_POINT_FILE_NAME
});
