// @modern-js/server-core@2.60.0/dist/cjs/adapters/node/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var node_exports = {};
__export(node_exports, {
  connectMid2HonoMid: () => import_hono.connectMid2HonoMid,
  createNodeServer: () => import_node.createNodeServer,
  createWebRequest: () => import_node.createWebRequest,
  getHtmlTemplates: () => import_plugins.getHtmlTemplates,
  getServerManifest: () => import_plugins.getServerManifest,
  httpCallBack2HonoMid: () => import_hono.httpCallBack2HonoMid,
  injectNodeSeverPlugin: () => import_plugins.injectNodeSeverPlugin,
  injectResourcePlugin: () => import_plugins.injectResourcePlugin,
  loadCacheConfig: () => import_helper.loadCacheConfig,
  loadServerCliConfig: () => import_helper.loadServerCliConfig,
  loadServerEnv: () => import_helper.loadServerEnv,
  loadServerPlugins: () => import_helper.loadServerPlugins,
  loadServerRuntimeConfig: () => import_helper.loadServerRuntimeConfig,
  sendResponse: () => import_node.sendResponse,
  serverStaticPlugin: () => import_plugins.serverStaticPlugin,
  writeReadableStreamToWritable: () => import_node.writeReadableStreamToWritable
});
module.exports = __toCommonJS(node_exports);
var import_hono = require("./hono");
var import_node = require("./node");
var import_plugins = require("./plugins");
var import_helper = require("./helper");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  connectMid2HonoMid,
  createNodeServer,
  createWebRequest,
  getHtmlTemplates,
  getServerManifest,
  httpCallBack2HonoMid,
  injectNodeSeverPlugin,
  injectResourcePlugin,
  loadCacheConfig,
  loadServerCliConfig,
  loadServerEnv,
  loadServerPlugins,
  loadServerRuntimeConfig,
  sendResponse,
  serverStaticPlugin,
  writeReadableStreamToWritable
});
