// @modern-js/plugin-garfish@2.60.0/dist/esm-node/runtime/loadable.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useCallback, useEffect, useState } from "react";
import { logger } from "../util";
const DEFAULT_LOADABLE = {
  delay: 200,
  timeout: 1e4,
  loading: null
};
function Loadable(WrapComponent) {
  return function(defaultLoadable) {
    return function Lodable(props) {
      var _props_loadable;
      const { loadable = defaultLoadable !== null && defaultLoadable !== void 0 ? defaultLoadable : DEFAULT_LOADABLE, ...otherProps } = props;
      let delayTimer = null;
      let timeoutTimer = null;
      const [state, setState] = useState(() => {
        const { delay, timeout } = loadable;
        const initState = {
          error: null,
          pastDelay: false,
          timedOut: false,
          isLoading: false
        };
        if (typeof delay === "number") {
          if (delay === 0) {
            initState.pastDelay = true;
          } else {
            delayTimer = setTimeout(() => {
              setStateWithMountCheck({
                pastDelay: true
              });
            }, delay);
          }
        }
        if (typeof timeout === "number") {
          timeoutTimer = setTimeout(() => {
            setStateWithMountCheck({
              timedOut: true
            });
          }, timeout);
        }
        return initState;
      });
      const LoadingComponent = (_props_loadable = props.loadable) === null || _props_loadable === void 0 ? void 0 : _props_loadable.loading;
      useEffect(() => {
        logger("Loadable render state", {
          state,
          props: otherProps,
          loadable,
          defaultLoadable
        });
        return () => {
          setStateWithMountCheck({
            isLoading: false,
            error: null
          });
          if (delayTimer) {
            clearTimeout(delayTimer);
            delayTimer = null;
          }
          if (timeoutTimer) {
            clearTimeout(timeoutTimer);
            timeoutTimer = null;
          }
        };
      }, []);
      const retry = useCallback(() => {
        setState({
          ...state,
          error: null,
          isLoading: true,
          timedOut: false
        });
      }, [
        state
      ]);
      const setStateWithMountCheck = useCallback((newState) => {
        setState((state2) => ({
          ...state2,
          ...newState
        }));
      }, [
        state
      ]);
      const showLoading = (state.isLoading || state.error) && LoadingComponent;
      return /* @__PURE__ */ _jsxs(_Fragment, {
        children: [
          showLoading && /* @__PURE__ */ _jsx(LoadingComponent, {
            isLoading: state.isLoading,
            pastDelay: state.pastDelay,
            timedOut: state.timedOut,
            error: state === null || state === void 0 ? void 0 : state.error,
            retry
          }),
          /* @__PURE__ */ _jsx(WrapComponent, {
            style: {
              display: showLoading ? "none" : "block"
            },
            setLoadingState: (props2) => {
              if (props2.error && !LoadingComponent) {
                throw props2.error;
              }
              setStateWithMountCheck(props2);
            },
            ...otherProps
          })
        ]
      });
    };
  };
}
export {
  Loadable
};
