// @modern-js/app-tools@2.60.0/dist/cjs/plugins/analyze/getServerRoutes.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var getServerRoutes_exports = {};
__export(getServerRoutes_exports, {
  getServerRoutes: () => getServerRoutes
});
module.exports = __toCommonJS(getServerRoutes_exports);
var import_fs = __toESM(require("fs"));
var import_path = __toESM(require("path"));
var import_utils = require("@modern-js/utils");
var import_routes = require("../../utils/routes");
var import_utils2 = require("./utils");
const applyBaseUrl = (baseUrl, routes) => {
  if (baseUrl) {
    if (Array.isArray(baseUrl)) {
      return baseUrl.reduce((previous, current) => [
        ...previous,
        ...applyBaseUrl(current, routes)
      ], []);
    } else {
      return routes.map((route) => {
        const urlPath = (0, import_utils.urlJoin)(baseUrl, route.urlPath);
        return {
          ...route,
          urlPath: urlPath === "/" ? urlPath : (0, import_utils.removeTailSlash)(urlPath)
        };
      });
    }
  }
  return routes;
};
const applyRouteOptions = (original, routeOptions) => {
  const { route, disableSpa } = routeOptions;
  original.isSPA = !disableSpa;
  !original.isSPA && (original.entryPath = import_path.default.dirname(original.entryPath));
  let routes;
  if (route) {
    if (Array.isArray(route)) {
      routes = route.map((url) => {
        if ((0, import_utils.isPlainObject)(url)) {
          const { path: urlPath, ...other } = url;
          return {
            ...original,
            ...other,
            urlPath
          };
        } else {
          return {
            ...original,
            urlPath: url
          };
        }
      });
    } else if ((0, import_utils.isPlainObject)(route)) {
      const { path: urlPath, ...other } = route;
      routes = [
        {
          ...original,
          ...other,
          urlPath
        }
      ];
    } else {
      routes = [
        {
          ...original,
          urlPath: route
        }
      ];
    }
  } else {
    routes = [
      original
    ];
  }
  return routes;
};
const collectHtmlRoutes = (entrypoints, appContext, config) => {
  var _deploy_worker;
  const { source: { mainEntryName }, html: { disableHtmlFolder }, output: { distPath: { html: htmlPath } = {} }, server: { baseUrl, routes, ssr, ssrByEntries }, deploy } = config;
  const { packageName } = appContext;
  const workerSSR = deploy === null || deploy === void 0 ? void 0 : (_deploy_worker = deploy.worker) === null || _deploy_worker === void 0 ? void 0 : _deploy_worker.ssr;
  let htmlRoutes = entrypoints.reduce((previous, { entryName }) => {
    const isMain = (0, import_routes.isMainEntry)(entryName, mainEntryName);
    const entryOptions = (0, import_utils.getEntryOptions)(entryName, isMain, ssr, ssrByEntries, packageName);
    const isSSR = Boolean(entryOptions);
    const isWorker = Boolean(workerSSR);
    const isStream = typeof entryOptions === "object" && entryOptions.mode === "stream";
    const { resHeaders } = (routes === null || routes === void 0 ? void 0 : routes[entryName]) || {};
    let route = {
      urlPath: `/${isMain ? "" : entryName}`,
      entryName,
      entryPath: (0, import_utils.removeLeadingSlash)(import_path.default.posix.normalize(`${htmlPath}/${entryName}${disableHtmlFolder ? ".html" : "/index.html"}`)),
      isSPA: true,
      isStream,
      isSSR,
      responseHeaders: resHeaders,
      worker: isWorker ? `${import_utils.SERVER_WORKER_BUNDLE_DIRECTORY}/${entryName}.js` : void 0,
      bundle: isSSR ? `${import_utils.SERVER_BUNDLE_DIRECTORY}/${entryName}.js` : void 0
    };
    if (routes === null || routes === void 0 ? void 0 : routes.hasOwnProperty(entryName)) {
      const routeOptions = (0, import_utils.isPlainObject)(routes[entryName]) ? routes[entryName] : {
        route: routes[entryName]
      };
      route = applyRouteOptions(route, routeOptions);
    }
    return Array.isArray(route) ? [
      ...previous,
      ...route
    ] : [
      ...previous,
      route
    ];
  }, []);
  htmlRoutes = applyBaseUrl(baseUrl, htmlRoutes);
  return htmlRoutes;
};
const collectStaticRoutes = (appContext, config) => {
  const { appDirectory } = appContext;
  const { source: { configDir }, server: { publicRoutes = {} } } = config;
  const publicFolder = import_path.default.resolve(appDirectory, configDir || "", "public");
  return import_fs.default.existsSync(publicFolder) ? (0, import_utils2.walkDirectory)(publicFolder).map((filePath) => {
    const urlPath = `${(0, import_utils.urlJoin)(toPosix(filePath).slice(toPosix(publicFolder).length))}`;
    return {
      urlPath: publicRoutes[(0, import_utils.removeLeadingSlash)(urlPath)] || urlPath,
      isSPA: true,
      isSSR: false,
      entryPath: toPosix(import_path.default.relative(import_path.default.resolve(appDirectory, configDir || ""), filePath))
    };
  }) : [];
};
const getServerRoutes = (entrypoints, { appContext, config }) => [
  ...collectHtmlRoutes(entrypoints, appContext, config),
  ...collectStaticRoutes(appContext, config)
];
const toPosix = (pathStr) => pathStr.split(import_path.default.sep).join(import_path.default.posix.sep);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getServerRoutes
});
