// @modern-js/app-tools@2.60.0/dist/esm-node/hooks.js
import { createAsyncWaterfall, createAsyncWorkflow, createParallelWorkflow } from "@modern-js/plugin";
const hooks = {
  _internalRuntimePlugins: createAsyncWaterfall(),
  modifyFileSystemRoutes: createAsyncWaterfall(),
  modifyServerRoutes: createAsyncWaterfall(),
  /** add entry point info to entrypoints array */
  modifyEntrypoints: createAsyncWaterfall(),
  /** add entry type */
  checkEntryPoint: createAsyncWaterfall(),
  generateEntryCode: createAsyncWorkflow(),
  htmlPartials: createAsyncWaterfall(),
  beforeGenerateRoutes: createAsyncWaterfall(),
  addDefineTypes: createAsyncWaterfall(),
  collectServerPlugins: createAsyncWaterfall(),
  _internalServerPlugins: createAsyncWaterfall(),
  beforeDev: createAsyncWorkflow(),
  afterDev: createAsyncWorkflow(),
  beforeCreateCompiler: createAsyncWorkflow(),
  afterCreateCompiler: createAsyncWorkflow(),
  beforePrintInstructions: createAsyncWaterfall(),
  beforeBuild: createAsyncWorkflow(),
  afterBuild: createAsyncWorkflow(),
  beforeDeploy: createAsyncWorkflow(),
  deploy: createAsyncWorkflow(),
  afterDeploy: createAsyncWorkflow(),
  beforeRestart: createAsyncWorkflow(),
  registerDev: createParallelWorkflow(),
  beforeDevTask: createParallelWorkflow(),
  registerBuildPlatform: createParallelWorkflow(),
  beforeBuildPlatform: createParallelWorkflow()
};
export {
  hooks
};
