// @modern-js/plugin-garfish@2.60.0/dist/esm/runtime/provider.js
import { _ as _async_to_generator } from "@swc/helpers/_/_async_to_generator";
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _ts_generator } from "@swc/helpers/_/_ts_generator";
import { jsx as _jsx } from "react/jsx-runtime";
import { render } from "@meta/runtime/browser";
import { createRoot } from "@meta/runtime/react";
import { createPortal, unmountComponentAtNode } from "react-dom";
function createProvider(id) {
  var _ref = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {}, customBootstrap = _ref.customBootstrap, beforeRender = _ref.beforeRender;
  return function(param) {
    var basename = param.basename, dom = param.dom;
    var root = null;
    return {
      render: function render1(param2) {
        var basename2 = param2.basename, dom2 = param2.dom, props = param2.props;
        return _async_to_generator(function() {
          var ModernRoot;
          return _ts_generator(this, function(_state) {
            switch (_state.label) {
              case 0:
                ModernRoot = createRoot(null);
                if (!customBootstrap)
                  return [
                    3,
                    2
                  ];
                return [
                  4,
                  customBootstrap(ModernRoot, function() {
                    return render(/* @__PURE__ */ _jsx(ModernRoot, _object_spread({
                      basename: basename2
                    }, props)), dom2);
                  })
                ];
              case 1:
                root = _state.sent();
                return [
                  3,
                  6
                ];
              case 2:
                if (!beforeRender)
                  return [
                    3,
                    4
                  ];
                return [
                  4,
                  beforeRender(ModernRoot, _object_spread({
                    basename: basename2
                  }, props))
                ];
              case 3:
                _state.sent();
                _state.label = 4;
              case 4:
                return [
                  4,
                  render(/* @__PURE__ */ _jsx(ModernRoot, _object_spread({
                    basename: basename2
                  }, props)), dom2)
                ];
              case 5:
                root = _state.sent();
                _state.label = 6;
              case 6:
                return [
                  2
                ];
            }
          });
        })();
      },
      destroy: function destroy(param2) {
        var dom2 = param2.dom;
        var node = dom2.querySelector("#".concat(id || "root")) || dom2;
        if (node) {
          if (process.env.IS_REACT18 === "true") {
            root.unmount();
          } else {
            unmountComponentAtNode(node);
          }
        }
      },
      // 兼容旧版本
      SubModuleComponent: function(props) {
        var ModernRoot = createRoot(null);
        return /* @__PURE__ */ createPortal(/* @__PURE__ */ _jsx(ModernRoot, _object_spread({
          basename
        }, props)), dom.querySelector("#".concat(id || "root")) || dom);
      },
      jupiter_submodule_app_key: function(props) {
        var ModernRoot = createRoot(null);
        return /* @__PURE__ */ createPortal(/* @__PURE__ */ _jsx(ModernRoot, _object_spread({
          basename
        }, props)), dom.querySelector("#".concat(id || "root")) || dom);
      }
    };
  };
}
export {
  createProvider
};
