// @modern-js/server-core@2.60.0/dist/cjs/plugins/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var plugins_exports = {};
__export(plugins_exports, {
  createDefaultPlugins: () => import_default.createDefaultPlugins,
  faviconPlugin: () => import_favicon.faviconPlugin,
  getRenderHandler: () => import_render.getRenderHandler,
  injectRenderHandlerPlugin: () => import_render.injectRenderHandlerPlugin,
  injectServerTiming: () => import_monitors.injectServerTiming,
  injectloggerPluigin: () => import_monitors.injectloggerPluigin,
  logPlugin: () => import_log.logPlugin,
  processedByPlugin: () => import_processedBy.processedByPlugin,
  renderPlugin: () => import_render.renderPlugin
});
module.exports = __toCommonJS(plugins_exports);
var import_render = require("./render");
var import_favicon = require("./favicon");
var import_monitors = require("./monitors");
var import_processedBy = require("./processedBy");
var import_log = require("./log");
var import_default = require("./default");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createDefaultPlugins,
  faviconPlugin,
  getRenderHandler,
  injectRenderHandlerPlugin,
  injectServerTiming,
  injectloggerPluigin,
  logPlugin,
  processedByPlugin,
  renderPlugin
});
