// @modern-js/app-tools@2.60.0/dist/esm/plugins/analyze/utils.js
import { _ as _async_to_generator } from "@swc/helpers/_/_async_to_generator";
import { _ as _to_consumable_array } from "@swc/helpers/_/_to_consumable_array";
import { _ as _ts_generator } from "@swc/helpers/_/_ts_generator";
import fs from "fs";
import path from "path";
import { JS_EXTENSIONS, getCommand, normalizeToPosixPath } from "@modern-js/utils";
import { parse } from "es-module-lexer";
import { transform } from "esbuild";
var walkDirectory = function(dir) {
  return fs.readdirSync(dir).reduce(function(previous, filename) {
    var filePath = path.join(dir, filename);
    if (fs.statSync(filePath).isDirectory()) {
      return _to_consumable_array(previous).concat(_to_consumable_array(walkDirectory(filePath)));
    } else {
      return _to_consumable_array(previous).concat([
        filePath
      ]);
    }
  }, []);
};
var replaceWithAlias = function(base, filePath, alias) {
  if (filePath.includes(base)) {
    return normalizeToPosixPath(path.join(alias, path.relative(base, filePath)));
  } else {
    return filePath;
  }
};
var parseModule = function() {
  var _ref = _async_to_generator(function(param) {
    var source, filename, content, result;
    return _ts_generator(this, function(_state) {
      switch (_state.label) {
        case 0:
          source = param.source, filename = param.filename;
          content = source;
          if (!JS_EXTENSIONS.some(function(ext) {
            return filename.endsWith(ext);
          }))
            return [
              3,
              2
            ];
          return [
            4,
            transform(content, {
              loader: path.extname(filename).slice(1),
              format: "esm"
            })
          ];
        case 1:
          result = _state.sent();
          content = result.code;
          _state.label = 2;
        case 2:
          return [
            4,
            parse(content)
          ];
        case 3:
          return [
            2,
            _state.sent()
          ];
      }
    });
  });
  return function parseModule2(_) {
    return _ref.apply(this, arguments);
  };
}();
var getServerCombinedModueFile = function(internalDirectory, entryName) {
  return path.join(internalDirectory, entryName, "server-loader-combined.js");
};
var checkIsBuildCommands = function() {
  var buildCommands = [
    "dev",
    "start",
    "build",
    "inspect",
    "deploy",
    "dev-worker"
  ];
  var command = getCommand();
  return buildCommands.includes(command);
};
var isSubDirOrEqual = function(parent, child) {
  if (parent === child) {
    return true;
  }
  var relative = path.relative(parent, child);
  var isSubdir = relative && !relative.startsWith("..") && !path.isAbsolute(relative);
  return Boolean(isSubdir);
};
export {
  checkIsBuildCommands,
  getServerCombinedModueFile,
  isSubDirOrEqual,
  parseModule,
  replaceWithAlias,
  walkDirectory
};
