// @modern-js/runtime@2.60.0/dist/cjs/core/server/string/loadable.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var loadable_exports = {};
__export(loadable_exports, {
  LoadableCollector: () => LoadableCollector
});
module.exports = __toCommonJS(loadable_exports);
var import_server = require("@loadable/server");
var import_utils = require("../utils");
const extname = (uri) => {
  if (typeof uri !== "string" || !uri.includes(".")) {
    return "";
  }
  return `.${uri === null || uri === void 0 ? void 0 : uri.split(".").pop()}` || "";
};
const generateChunks = (chunks, ext) => chunks.filter((chunk) => Boolean(chunk.url)).filter((chunk) => extname(chunk.url).slice(1) === ext);
const checkIsInline = (chunk, enableInline) => {
  if (process.env.NODE_ENV === "production") {
    if (enableInline instanceof RegExp) {
      return enableInline.test(chunk.url);
    } else {
      return Boolean(enableInline);
    }
  } else {
    return false;
  }
};
const readAsset = async (chunk) => {
  const fs = await Promise.resolve().then(() => __toESM(require("fs/promises")));
  const path = await Promise.resolve().then(() => __toESM(require("path")));
  const filepath = path.resolve(__dirname, chunk.filename);
  return fs.readFile(filepath, "utf-8");
};
class LoadableCollector {
  get existsAssets() {
    var _routeManifest_routeAssets_entryName, _routeManifest_routeAssets;
    const { routeManifest, entryName } = this.options;
    return routeManifest === null || routeManifest === void 0 ? void 0 : (_routeManifest_routeAssets = routeManifest.routeAssets) === null || _routeManifest_routeAssets === void 0 ? void 0 : (_routeManifest_routeAssets_entryName = _routeManifest_routeAssets[entryName]) === null || _routeManifest_routeAssets_entryName === void 0 ? void 0 : _routeManifest_routeAssets_entryName.assets;
  }
  collect(comopnent) {
    const { stats, entryName } = this.options;
    if (!stats) {
      return comopnent;
    }
    this.extractor = new import_server.ChunkExtractor({
      stats,
      entrypoints: [
        entryName
      ]
    });
    return this.extractor.collectChunks(comopnent);
  }
  async effect() {
    if (!this.extractor) {
      return;
    }
    const { extractor } = this;
    const chunks = extractor.getChunkAssets(extractor.chunks);
    const scriptChunks = generateChunks(chunks, "js");
    const styleChunks = generateChunks(chunks, "css");
    this.emitLoadableScripts(extractor);
    await this.emitScriptAssets(scriptChunks);
    await this.emitStyleAssets(styleChunks);
  }
  emitLoadableScripts(extractor) {
    const check = (scripts2) => (scripts2 || "").includes("__LOADABLE_REQUIRED_CHUNKS___ext");
    const scripts = extractor.getScriptTags();
    if (!check(scripts)) {
      return;
    }
    const { chunkSet } = this.options;
    const s = scripts.split("</script>").slice(0, 2).map((i) => `${i}</script>`).join("");
    chunkSet.jsChunk += s;
  }
  async emitScriptAssets(chunks) {
    const { template, nonce, chunkSet, config } = this.options;
    const { scriptLoading = "defer", enableInlineScripts } = config;
    const scriptLoadingAtr = {
      defer: scriptLoading === "defer" ? true : void 0,
      type: scriptLoading === "module" ? "module" : void 0,
      async: scriptLoading === "async" ? true : void 0
    };
    const attributes = (0, import_utils.attributesToString)(this.generateAttributes({
      nonce,
      ...scriptLoadingAtr
    }));
    const scripts = await Promise.all(chunks.filter((chunk) => {
      var _this_existsAssets;
      const jsChunkReg = new RegExp(`<script .*src="${chunk.url}".*>`);
      return !jsChunkReg.test(template) && !((_this_existsAssets = this.existsAssets) === null || _this_existsAssets === void 0 ? void 0 : _this_existsAssets.includes(chunk.path));
    }).map(async (chunk) => {
      const script = `<script${attributes} src="${chunk.url}"></script>`;
      if ((0, import_utils.checkIsNode)() && checkIsInline(chunk, enableInlineScripts)) {
        return readAsset(chunk).then((content) => `<script>${content}</script>`).catch((_) => {
          return script;
        });
      } else {
        return script;
      }
    }));
    chunkSet.jsChunk += scripts.filter((script) => Boolean(script)).join("");
  }
  async emitStyleAssets(chunks) {
    const { template, chunkSet, config } = this.options;
    const { enableInlineStyles } = config;
    const atrributes = (0, import_utils.attributesToString)(this.generateAttributes());
    const css = await Promise.all(chunks.filter((chunk) => {
      var _this_existsAssets;
      const cssChunkReg = new RegExp(`<link .*href="${chunk.url}".*>`);
      return !cssChunkReg.test(template) && !((_this_existsAssets = this.existsAssets) === null || _this_existsAssets === void 0 ? void 0 : _this_existsAssets.includes(chunk.path));
    }).map(async (chunk) => {
      const link = `<link${atrributes} href="${chunk.url}" rel="stylesheet" />`;
      if ((0, import_utils.checkIsNode)() && checkIsInline(chunk, enableInlineStyles)) {
        return readAsset(chunk).then((content) => `<style>${content}</style>`).catch((_) => {
          return link;
        });
      } else {
        return link;
      }
    }));
    chunkSet.cssChunk += css.filter((css2) => Boolean(css2)).join("");
  }
  generateAttributes(extraAtr = {}) {
    const { config } = this.options;
    const { crossorigin } = config;
    const attributes = {};
    if (crossorigin) {
      attributes.crossorigin = crossorigin === true ? "anonymous" : crossorigin;
    }
    return {
      ...attributes,
      ...extraAtr
    };
  }
  constructor(options) {
    this.options = options;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LoadableCollector
});
