// @modern-js/app-tools@2.60.0/dist/cjs/hooks.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var hooks_exports = {};
__export(hooks_exports, {
  hooks: () => hooks
});
module.exports = __toCommonJS(hooks_exports);
var import_plugin = require("@modern-js/plugin");
const hooks = {
  _internalRuntimePlugins: (0, import_plugin.createAsyncWaterfall)(),
  modifyFileSystemRoutes: (0, import_plugin.createAsyncWaterfall)(),
  modifyServerRoutes: (0, import_plugin.createAsyncWaterfall)(),
  /** add entry point info to entrypoints array */
  modifyEntrypoints: (0, import_plugin.createAsyncWaterfall)(),
  /** add entry type */
  checkEntryPoint: (0, import_plugin.createAsyncWaterfall)(),
  generateEntryCode: (0, import_plugin.createAsyncWorkflow)(),
  htmlPartials: (0, import_plugin.createAsyncWaterfall)(),
  beforeGenerateRoutes: (0, import_plugin.createAsyncWaterfall)(),
  addDefineTypes: (0, import_plugin.createAsyncWaterfall)(),
  collectServerPlugins: (0, import_plugin.createAsyncWaterfall)(),
  _internalServerPlugins: (0, import_plugin.createAsyncWaterfall)(),
  beforeDev: (0, import_plugin.createAsyncWorkflow)(),
  afterDev: (0, import_plugin.createAsyncWorkflow)(),
  beforeCreateCompiler: (0, import_plugin.createAsyncWorkflow)(),
  afterCreateCompiler: (0, import_plugin.createAsyncWorkflow)(),
  beforePrintInstructions: (0, import_plugin.createAsyncWaterfall)(),
  beforeBuild: (0, import_plugin.createAsyncWorkflow)(),
  afterBuild: (0, import_plugin.createAsyncWorkflow)(),
  beforeDeploy: (0, import_plugin.createAsyncWorkflow)(),
  deploy: (0, import_plugin.createAsyncWorkflow)(),
  afterDeploy: (0, import_plugin.createAsyncWorkflow)(),
  beforeRestart: (0, import_plugin.createAsyncWorkflow)(),
  registerDev: (0, import_plugin.createParallelWorkflow)(),
  beforeDevTask: (0, import_plugin.createParallelWorkflow)(),
  registerBuildPlatform: (0, import_plugin.createParallelWorkflow)(),
  beforeBuildPlatform: (0, import_plugin.createParallelWorkflow)()
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  hooks
});
