// @modern-js/runtime@2.60.0/dist/esm/router/cli/constants.js
var PAGES_DIR_NAME = "pages";
var NESTED_ROUTES_DIR = "routes";
var FILE_SYSTEM_ROUTES_FILE_NAME = "routes.js";
var LOADER_EXPORT_NAME = "loader";
var ACTION_EXPORT_NAME = "action";
var TEMP_LOADERS_DIR = "__loaders__";
var FILE_SYSTEM_ROUTES_LAYOUT = "_layout";
var FILE_SYSTEM_ROUTES_GLOBAL_LAYOUT = "_app";
var FILE_SYSTEM_ROUTES_INDEX = "index";
var FILE_SYSTEM_ROUTES_IGNORED_REGEX = /\.(d|test|spec|e2e)\.(js|jsx|ts|tsx)$/;
var FILE_SYSTEM_ROUTES_DYNAMIC_REGEXP = /^\[(\S+)\]([*+?]?)$/;
var FILE_SYSTEM_ROUTES_COMPONENTS_DIR = "internal_components";
var APP_CONFIG_NAME = "config";
var APP_INIT_EXPORTED = "init";
var NESTED_ROUTE = {
  LAYOUT_FILE: "layout",
  LAYOUT_CONFIG_FILE: "layout.config",
  LAYOUT_LOADER_FILE: "layout.loader",
  LAYOUT_DATA_FILE: "layout.data",
  LAYOUT_CLIENT_LOADER: "layout.data.client",
  PAGE_FILE: "page",
  PAGE_CONFIG_FILE: "page.config",
  PAGE_LOADER_FILE: "page.loader",
  PAGE_DATA_FILE: "page.data",
  PAGE_CLIENT_LOADER: "page.data.client",
  SPLATE_FILE: "$",
  SPLATE_CONFIG_FILE: "$.config",
  SPLATE_LOADER_FILE: "$.loader",
  SPLATE_DATA_FILE: "$.data",
  SPLATE_CLIENT_DATA: "$.data.client",
  LOADING_FILE: "loading",
  ERROR_FILE: "error",
  LOADER_FILE: "loader"
};
export {
  ACTION_EXPORT_NAME,
  APP_CONFIG_NAME,
  APP_INIT_EXPORTED,
  FILE_SYSTEM_ROUTES_COMPONENTS_DIR,
  FILE_SYSTEM_ROUTES_DYNAMIC_REGEXP,
  FILE_SYSTEM_ROUTES_FILE_NAME,
  FILE_SYSTEM_ROUTES_GLOBAL_LAYOUT,
  FILE_SYSTEM_ROUTES_IGNORED_REGEX,
  FILE_SYSTEM_ROUTES_INDEX,
  FILE_SYSTEM_ROUTES_LAYOUT,
  LOADER_EXPORT_NAME,
  NESTED_ROUTE,
  NESTED_ROUTES_DIR,
  PAGES_DIR_NAME,
  TEMP_LOADERS_DIR
};
