// @modern-js/runtime@2.60.0/dist/cjs/router/cli/constants.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var constants_exports = {};
__export(constants_exports, {
  ACTION_EXPORT_NAME: () => ACTION_EXPORT_NAME,
  APP_CONFIG_NAME: () => APP_CONFIG_NAME,
  APP_INIT_EXPORTED: () => APP_INIT_EXPORTED,
  FILE_SYSTEM_ROUTES_COMPONENTS_DIR: () => FILE_SYSTEM_ROUTES_COMPONENTS_DIR,
  FILE_SYSTEM_ROUTES_DYNAMIC_REGEXP: () => FILE_SYSTEM_ROUTES_DYNAMIC_REGEXP,
  FILE_SYSTEM_ROUTES_FILE_NAME: () => FILE_SYSTEM_ROUTES_FILE_NAME,
  FILE_SYSTEM_ROUTES_GLOBAL_LAYOUT: () => FILE_SYSTEM_ROUTES_GLOBAL_LAYOUT,
  FILE_SYSTEM_ROUTES_IGNORED_REGEX: () => FILE_SYSTEM_ROUTES_IGNORED_REGEX,
  FILE_SYSTEM_ROUTES_INDEX: () => FILE_SYSTEM_ROUTES_INDEX,
  FILE_SYSTEM_ROUTES_LAYOUT: () => FILE_SYSTEM_ROUTES_LAYOUT,
  LOADER_EXPORT_NAME: () => LOADER_EXPORT_NAME,
  NESTED_ROUTE: () => NESTED_ROUTE,
  NESTED_ROUTES_DIR: () => NESTED_ROUTES_DIR,
  PAGES_DIR_NAME: () => PAGES_DIR_NAME,
  TEMP_LOADERS_DIR: () => TEMP_LOADERS_DIR
});
module.exports = __toCommonJS(constants_exports);
const PAGES_DIR_NAME = "pages";
const NESTED_ROUTES_DIR = "routes";
const FILE_SYSTEM_ROUTES_FILE_NAME = "routes.js";
const LOADER_EXPORT_NAME = "loader";
const ACTION_EXPORT_NAME = "action";
const TEMP_LOADERS_DIR = "__loaders__";
const FILE_SYSTEM_ROUTES_LAYOUT = "_layout";
const FILE_SYSTEM_ROUTES_GLOBAL_LAYOUT = "_app";
const FILE_SYSTEM_ROUTES_INDEX = "index";
const FILE_SYSTEM_ROUTES_IGNORED_REGEX = /\.(d|test|spec|e2e)\.(js|jsx|ts|tsx)$/;
const FILE_SYSTEM_ROUTES_DYNAMIC_REGEXP = /^\[(\S+)\]([*+?]?)$/;
const FILE_SYSTEM_ROUTES_COMPONENTS_DIR = "internal_components";
const APP_CONFIG_NAME = "config";
const APP_INIT_EXPORTED = "init";
const NESTED_ROUTE = {
  LAYOUT_FILE: "layout",
  LAYOUT_CONFIG_FILE: "layout.config",
  LAYOUT_LOADER_FILE: "layout.loader",
  LAYOUT_DATA_FILE: "layout.data",
  LAYOUT_CLIENT_LOADER: "layout.data.client",
  PAGE_FILE: "page",
  PAGE_CONFIG_FILE: "page.config",
  PAGE_LOADER_FILE: "page.loader",
  PAGE_DATA_FILE: "page.data",
  PAGE_CLIENT_LOADER: "page.data.client",
  SPLATE_FILE: "$",
  SPLATE_CONFIG_FILE: "$.config",
  SPLATE_LOADER_FILE: "$.loader",
  SPLATE_DATA_FILE: "$.data",
  SPLATE_CLIENT_DATA: "$.data.client",
  LOADING_FILE: "loading",
  ERROR_FILE: "error",
  LOADER_FILE: "loader"
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ACTION_EXPORT_NAME,
  APP_CONFIG_NAME,
  APP_INIT_EXPORTED,
  FILE_SYSTEM_ROUTES_COMPONENTS_DIR,
  FILE_SYSTEM_ROUTES_DYNAMIC_REGEXP,
  FILE_SYSTEM_ROUTES_FILE_NAME,
  FILE_SYSTEM_ROUTES_GLOBAL_LAYOUT,
  FILE_SYSTEM_ROUTES_IGNORED_REGEX,
  FILE_SYSTEM_ROUTES_INDEX,
  FILE_SYSTEM_ROUTES_LAYOUT,
  LOADER_EXPORT_NAME,
  NESTED_ROUTE,
  NESTED_ROUTES_DIR,
  PAGES_DIR_NAME,
  TEMP_LOADERS_DIR
});
